translate/compiler.go

```diff
@@ -37,19 +44,102 @@ func CompileContract(c ast.Contract) (Bytecode, error) {
 
 	memTracer := NewMemEntryTable()
 
+	funcMap := FuncMap{}
+
 	for _, f := range c.Functions {
+		funcMap.Declare(*f, *bytecode)
+
 		if err := compileFunction(*f, bytecode, memTracer); err != nil {
 			return *bytecode, err
 		}
 	}
 
-	if err := generateFuncJumper(bytecode); err != nil {
+	if err := expectFuncJumper(c, funcMap); err != nil {
+		return *bytecode, err
+	}
+
+	if err := generateFuncJumper(c, bytecode, funcMap); err != nil {
 		return *bytecode, err
 	}
 
 	return *bytecode, nil
 }
 
+// Expects a size of the function jumper and updates the function map.
+func expectFuncJumper(c ast.Contract, funcMap FuncMap) error {
+	expectedJpr := NewBytecode()
+	if err := generateFuncJumper(c, expectedJpr, funcMap); err != nil {
+		return err
+	}
+
+	// Updates the function map with a size of function jumper.
+	for funcSel, pc := range funcMap {
+		funcMap[funcSel] = pc + len(expectedJpr.RawByte)
+	}
+
+	return nil
+}
+
+// Generates the function jumper in front of the bytecode.
+func generateFuncJumper(c ast.Contract, bytecode *Bytecode, funcMap FuncMap) error {
+	funcJpr := NewBytecode()
+
+	// Loads the function selector of call function.
+	funcJpr.Emerge(opcode.LoadFunc)
+
+	// Compares and finds the corresponding function selector.
+	for _, f := range c.Functions {
+		if err := compileFunctionLiteral(f, funcJpr, funcMap); err != nil {
+			return err
+		}
+	}
+
+	// No match to any function selector, Revert!
+	funcJpr.Emerge(opcode.JumpDst)
+	operand, err := encoding.EncodeOperand(0)
+	if err != nil {
+		return err
+	}
+	funcJpr.Emerge(opcode.Push, operand)
+	funcJpr.Emerge(opcode.DUP)
+	funcJpr.Emerge(opcode.Returning)
+
+	// Add function jumper to bytecode
+	// [bytecode] => [function jumper][bytecode]
+	funcJpr.Append(bytecode)
+	bytecode = funcJpr
+
+	return nil
+}
+
+// TODO: implement me w/ test cases :-)
+func compileFunctionLiteral(s *ast.FunctionLiteral, bytecode *Bytecode, funcMap FuncMap) error {
```

Comment:
`ast.FunctionLiteral` contains not only `Name`, `Parameters` but also `Body`. so I thought that `compileFunctonLiteral` was compiling entire function. 

But this is about compiling one of function jumper switch case. So little bit confusing

Reply:
`Body` in `ast.FunctionLiteral` would be compiled in `compileFunction()`, not `compileFunctionLiteral()`. `compileFunctionLiteral()` just get the function selector and add a logic to find the function location called. I think `compileFunctionLiteral()` name is so confusing. How about `compileFuncSelector()`? If you have any idea, please recommend it!

Reply:
I think compileFuncSelector() is good.